src/vivarium_census_prl_synth_pop/components/observers.py

```diff
@@ -494,3 +472,191 @@ def get_observation(self, event: Event) -> pd.DataFrame:
+
+    @property
+    def name(self):
+        return f"tax_w2_observer"
+
+    @property
+    def input_columns(self):
+        return self.DEFAULT_INPUT_COLUMNS + self.ADDITIONAL_INPUT_COLUMNS
+
+    @property
+    def input_values(self):
+        return self.INPUT_VALUES
+
+    @property
+    def output_columns(self):
+        return self.OUTPUT_COLUMNS
+
+    def setup(self, builder: Builder):
+        super().setup(builder)
+        self.clock = builder.time.clock()
+
+        builder.event.register_listener("time_step", self.on_time_step)
+        self.income_to_date = empty_income_series()
+        self.time_step = builder.configuration.time.step_size  # in days
+
+    def on_time_step(self, event):
+        pop = self.population_view.get(
+            event.index,
+        )
+        pop["income"] = self.pipelines["income"](pop.index)
+
+        # increment income for all person/employment pairs with income > 0
+        income_this_time_step = pd.Series(
+            pop["income"].values * self.time_step / DAYS_PER_YEAR,
+            index=pd.MultiIndex.from_arrays(
+                [pop.index, pop["employer_id"]], names=["simulant_id", "employer_id"]
+            ),
+        )
+
+        income_this_time_step = income_this_time_step[income_this_time_step > 0]
+
+        self.income_to_date = self.income_to_date.add(income_this_time_step, fill_value=0.0)
+
+    def to_observe(self, event: Event) -> bool:
+        """Only observe if this is the final time step of the sim"""
+        tax_date = dt.datetime(event.time.year, 1, 1)
+        return self.clock() < tax_date <= event.time
+
+    def get_observation(self, event: Event) -> pd.DataFrame:
+        pop = self.population_view.get(
+            event.index,
+        )
+        pop["income"] = self.pipelines["income"](pop.index)
+
+        ### create dataframe of all person/employment pairs
+
+        # start with income to date, which has simulant_id and employer_id as multi-index
+        self.income_to_date.name = "income"  # HACK: it would be nice if this name stayed
+        # with the pd.Series, but it is getting lost at some point in the computation
+
+        df_w2 = self.income_to_date.reset_index()
+
+        # merge in simulant columns based on simulant id
+        for col in [
+            "first_name_id",
+            "middle_name_id",
+            "last_name_id",
+            "age",
+            "date_of_birth",
+            "sex",
+            "ssn",
+            "address_id",
+            "housing_type",
+        ]:
+            df_w2[col] = df_w2["simulant_id"].map(pop[col])
+
+        # for simulants without ssn, record a simulant_id for a random household
+        # member with an ssn, if one exists
+
+        simulants_wo_ssn = pd.Series(
+            df_w2[~df_w2["ssn"]].index, index=df_w2[~df_w2["ssn"]].index
+        )
+        household_members_w_ssn = (
+            pop[pop["ssn"]].groupby("address_id").apply(lambda df_g: list(df_g.index))
```

Comment:
You can do this without an apply and end up with a Series with the address id as an index with 
```
pd.Series(pop[pop["ssn"]], index=pop.loc[pop["ssn"], "address_id"]
```

Reply:
I want something a little complicated here, to get a series of lists:

```
(Pdb++) pop[pop["ssn"]].groupby("address_id").apply(lambda df_g: list(df_g.index))
address_id
1.0      [1, 2, 3]
2.0            [4]
3.0         [5, 6]
4.0      [7, 8, 9]
5.0       [10, 11]
           ...
913.0        [302]
914.0        [486]
915.0        [600]
916.0        [937]
917.0       [1197]
Length: 859, dtype: object
```

I don't know how to make your proposed code achieve this.

Reply:
I don't know exactly what pop looks like at this point (`pop[pop["ssn"]]` is freaking me out), but this is still an apply but a bit cleaner:

```
pop[pop["ssn"]].groupby("address_id").apply(list)
```

Reply:
oh, cool simplification of the `.apply`!

I could switch to `pop.query("ssn == True")` if you find that clearer.  I like that a little better, but I thought `pop[pop["ssn"]]` was closer to our style.

Reply:
shoot, the `.apply(list)` doesn't work:

```
(Pdb++) pop[pop["ssn"]].groupby("address_id").apply(lambda df_g: list(df_g.index))
address_id
1.0          [2, 3]
2.0             [4]
3.0          [5, 6]
4.0       [7, 8, 9]
5.0            [10]
            ...
8934.0      [17681]
8935.0      [19060]
8936.0      [19642]
8938.0      [19825]
8939.0      [19862]
Length: 8224, dtype: object


(Pdb++) pop[pop["ssn"]].groupby("address_id").apply(list)
address_id
1.0       [first_name_id, middle_name_id, last_name_id, ...
2.0       [first_name_id, middle_name_id, last_name_id, ...
3.0       [first_name_id, middle_name_id, last_name_id, ...
4.0       [first_name_id, middle_name_id, last_name_id, ...
5.0       [first_name_id, middle_name_id, last_name_id, ...
                                ...
8934.0    [first_name_id, middle_name_id, last_name_id, ...
8935.0    [first_name_id, middle_name_id, last_name_id, ...
8936.0    [first_name_id, middle_name_id, last_name_id, ...
8938.0    [first_name_id, middle_name_id, last_name_id, ...
8939.0    [first_name_id, middle_name_id, last_name_id, ...
Length: 8224, dtype: object
```

Reply:
@aflaxman I think what Steve was going for was `pop[pop["ssn"]].reset_index().groupby("address_id")["index"].apply(list)`.

Reply:
oh, cool!  Does that work for you?

I get
```
(Pdb++) pop[pop["ssn"]].groupby("address_id")["index"].apply(list)
*** KeyError: 'Column not found: index'
Traceback (most recent call last):
  File "/homes/abie/.conda/envs/vivarium_census_prl_synth_pop_2/lib/python3.8/site-packages/pandas/core/groupby/generic.py", line 1338, in __getitem__
    return super().__getitem__(key)
  File "/homes/abie/.conda/envs/vivarium_census_prl_synth_pop_2/lib/python3.8/site-packages/pandas/core/base.py", line 250, in __getitem__
    raise KeyError(f"Column not found: {key}")
```
when I try it.

Reply:
@aflaxman You need the `.reset_index()` in my comment above.